### PR TITLE
Fix for ch4/ofi: RMA key allocator rework

### PR DESCRIPTION
**Pull request by yfguo**
I use this to test my fix for pmodels/mpich#3493 since I cannot push to Alex's PR.

### CODE REVIEW

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
The results looks good now and I have force pushed some cleanup.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 1]
```diff
@@ -333,6 +333,16 @@ uint64_t MPIDI_OFI_index_allocator_alloc(void *_indexmap, MPL_memory_class class
 void MPIDI_OFI_index_allocator_free(void *_indexmap, uint64_t index);
 void MPIDI_OFI_index_allocator_destroy(void *_indexmap);
 
+static inline int MPIDI_OFI_index_is_valid(uint64_t index, unsigned index_bitsize)
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: `MPL_STATIC_INLINE_PREFIX`

Comment (hajimefu) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Let's use `bool` as the return value.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 2]
```diff
@@ -333,6 +333,16 @@ uint64_t MPIDI_OFI_index_allocator_alloc(void *_indexmap, MPL_memory_class class
...
+{
+    MPIR_Assert(index_bitsize <= 64);
+    if (index_bitsize < 64) {
+        return index < (((uint64_t) 1) << index_bitsize);
+    } else {
+        return 1;
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: `return true;`

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 1 above)
Comment (hajimefu) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Since `index_bitsize` is always `MPIDI_Global.max_mr_key_size * 8`, how about drop this parameter from the function and refer to `MPIDI_Global.max_mr_key_size * 8` internally?

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 3]
```diff
@@ -388,6 +341,7 @@ typedef struct {
     /* Window/RMA Globals */
     void *win_map;
     uint64_t rma_issued_cntr;
+    void *mr_key_allocator;
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_types.h]: In #3493 I commented below:
>Also, if I understood correctly, this single allocator is now used for
> * generating window ID for RMA (=equal to MR key for RMA)
> * generating MR key for pt2pt LMT
> These design choices have to be written somewhere, probably here.

I think we need more documentation like this as a comment as well as commit message. @yfguo @alexander-sannikov

File: src/mpid/ch4/netmod/ofi/ofi_comm.h [hunk 4]
```diff
@@ -28,8 +28,8 @@ static inline int MPIDI_NM_mpi_comm_create_hook(MPIR_Comm * comm)
 
     MPIDI_CH4U_map_create(&MPIDI_OFI_COMM(comm).huge_send_counters, MPL_MEM_COMM);
     MPIDI_CH4U_map_create(&MPIDI_OFI_COMM(comm).huge_recv_counters, MPL_MEM_COMM);
-    MPIDI_OFI_index_allocator_create(&MPIDI_OFI_COMM(comm).win_id_allocator, 0, MPL_MEM_RMA);
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_comm.h]: Removing the `start` param should be in another patch.

File: src/mpid/ch4/netmod/ofi/ofi_types.h (hunk 3 above)
Comment (hajimefu) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: Commit: "per-comm MR-key allocator is replaced with global one"

The commit message totally lacks *why* we had to do replacement.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 1 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: I can change the return type to `bool`. I thought about passing the size as bytes rather than bits (the current), but I think the function is more general.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
I have cleaned up the commits and added the explanation of the design choice based on the discussion on Slack. If the tests run good, it is ready to be merged.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 5]
```diff
@@ -327,7 +327,7 @@ int MPIDI_OFI_control_handler(int handler_id, void *am_hdr,
                               MPIDIG_am_target_cmpl_cb * target_cmpl_cb, MPIR_Request ** req);
 int MPIDI_OFI_control_dispatch(void *buf);
 void MPIDI_OFI_index_datatypes(void);
-void MPIDI_OFI_index_allocator_create(void **_indexmap, int start, MPL_memory_class class);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Why is this no longer needed? Please explain in the commit message.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h [hunk 6]
```diff
@@ -342,7 +342,7 @@ static inline int MPIDI_OFI_handle_lmt_ack(MPIDI_OFI_am_header_t * msg_hdr)
         MPIR_Assert(MPIDI_OFI_KEY_TYPE_HUGE_RMA == key_type);
 
         MPIDI_OFI_index_allocator_free(MPIDI_OFI_COMM(MPIR_Process.comm_world).rma_id_allocator,
-                                       idx_back);
+                                       (uint64_t) idx_back);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_am_events.h]: Why are we casting here? Are we always truncating the allocated key and casting it back? That seems weird.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h [hunk 7]
```diff
@@ -341,8 +341,7 @@ static inline int MPIDI_OFI_handle_lmt_ack(MPIDI_OFI_am_header_t * msg_hdr)
         MPIDI_OFI_rma_key_unpack(mr_key, NULL, &key_type, &idx_back);
         MPIR_Assert(MPIDI_OFI_KEY_TYPE_HUGE_RMA == key_type);
 
-        MPIDI_OFI_index_allocator_free(MPIDI_OFI_COMM(MPIR_Process.comm_world).rma_id_allocator,
-                                       (uint64_t) idx_back);
+        MPIDI_OFI_index_allocator_free(MPIDI_Global.mr_key_allocator, (uint64_t) idx_back);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_am_events.h]: Will we have multiple allocators at some point? I'm wondering if passing this global var through the API is really necessary.

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 8]
```diff
@@ -67,53 +67,6 @@
 #endif
 #endif
 
-/* RMA Key Space division
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_types.h]: What does the key contain now?

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h [hunk 9]
```diff
@@ -299,9 +299,9 @@ static inline int MPIDI_OFI_am_isend_long(int rank,
          * is not available here */
         lmt_info->rma_key =
             MPIDI_OFI_index_allocator_alloc(MPIDI_Global.mr_key_allocator, MPL_MEM_RMA);
-        MPIR_ERR_CHKANDSTMT(lmt_info->rma_key >=
-                            (((uint64_t) 1) << (MPIDI_Global.max_mr_key_size * 8)), mpi_errno,
-                            MPI_ERR_OTHER, goto fn_fail, "**ofid_mr_reg");
+        MPIR_ERR_CHKANDSTMT(!MPIDI_OFI_index_is_valid(lmt_info->rma_key,
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: Same comment about passing global var through the API.

File: src/mpid/ch4/netmod/ofi/ofi_types.h (hunk 8 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: Now is just a unique number. It does not have anything like key type or context ID.

**Comment by minsii** [comment]
Just want to let you know that I did not get a chance to review the PR by myself. But I will also need the documentation as required by @hajimefu . I can review it tomorrow if needed.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 6 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: Hmm, I think I need to take a another look of the implementation.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 10]
```diff
@@ -327,9 +327,10 @@ int MPIDI_OFI_control_handler(int handler_id, void *am_hdr,
                               MPIDIG_am_target_cmpl_cb * target_cmpl_cb, MPIR_Request ** req);
 int MPIDI_OFI_control_dispatch(void *buf);
 void MPIDI_OFI_index_datatypes(void);
+
 void MPIDI_OFI_index_allocator_create(void **_indexmap, MPL_memory_class class);
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Here is an empty line change. Can you please remove it ?

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h [hunk 11]
```diff
@@ -297,11 +297,11 @@ static inline int MPIDI_OFI_am_isend_long(int rank,
...
-        uint64_t index =
+        lmt_info->rma_key =
             MPIDI_OFI_index_allocator_alloc(MPIDI_Global.mr_key_allocator, MPL_MEM_RMA);
-        MPIR_Assert(index < MPIDI_Global.max_huge_rmas);
-        lmt_info->rma_key = MPIDI_OFI_rma_key_pack(comm->context_id,
-                                                   MPIDI_OFI_KEY_TYPE_HUGE_RMA, index);
+        MPIR_ERR_CHKANDSTMT(lmt_info->rma_key >=
+                            (((uint64_t) 1) << (MPIDI_Global.max_mr_key_size * 8)), mpi_errno,
+                            MPI_ERR_OTHER, goto fn_fail, "**ofid_mr_reg");
     } else {
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: I do not understand the condition check in `MPIR_ERR_CHKANDSTMT`. Can we have a short comment here explaining the logic ?

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 12]
```diff
@@ -607,22 +607,7 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
...
-        MPIDI_Global.context_shift = MPIDI_OFI_CONTEXT_SHIFT_32;
-        MPIDI_Global.rma_key_type_bits = MPIDI_OFI_MAX_KEY_TYPE_BITS_32;
-    } else if (MPIDI_Global.max_mr_key_size >= 2) {
-        MPIDI_Global.max_rma_key_bits = MPIDI_OFI_MAX_KEY_BITS_16;
-        MPIDI_Global.max_huge_rmas = MPIDI_OFI_MAX_HUGE_RMAS_16;
-        MPIDI_Global.context_shift = MPIDI_OFI_CONTEXT_SHIFT_16;
-        MPIDI_Global.rma_key_type_bits = MPIDI_OFI_MAX_KEY_TYPE_BITS_16;
-    } else {
+    if (MPIDI_Global.max_mr_key_size < 2) {
         MPIR_ERR_SETFATALANDJUMP4(mpi_errno,
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: Can we briefly explain why error should be reported here ?

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h (hunk 11 above)
Comment (minsii) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: It looks like this commit just fixes the previous commit . Can we squash these two commits ?

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 7 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: I think the index allocator is a general implementation that does not tie to MR key itself. Therefore, it is better to keep passing this global allocator and keep the implementation reusable.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 6 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: The casting here is just suppressing the warning. Even without the casting, the compiler would do an implicit casting.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 5 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: I am not sure why it was designed like this in the first place. My only guess is because we have two allocators and we want to avoid id collisions from the two allocators. But, that will be weird since we only offset the output of these two allocators by one.

Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: The simplified implementation no longer needs that since we are allocating from the same allocator which won't have any collision.

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h (hunk 9 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: Answered above.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 10 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Will do.

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h (hunk 11 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: The provider may support a key less than 64 bits. Since the id allocator only allocates 64-bit keys (sequentially started from 0), we need to check if the allocated key is within the supported key length. For example, if the provider supports 32-bit key, then any key larger than 0xFFFF is invalid.

Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: I will squash them.

Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: It is poor practice to simply abort with a valid provider, unless it's documented as a required OFI feature by MPICH.  @raffenet was putting together a list.  It'll need to go in that list if it's really a required feature.

Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: 1. The allocator *can* allocate 64-bit keys, but we know at init time what the maximum key size is in `MPIDI_Global.max_mr_key_size`. Are we allowing the allocator to produce keys that can't be used?

2. libfabric will return errors for bad keys.
```-FI_ENOKEY
The requested_key is already in use.
-FI_EKEYREJECTED
The requested_key is not available. They key may be out of the range supported by the provider, or the provider may not support user-requested memory registration keys.
```
It seems to me this check can just be removed.

File: src/mpid/ch4/netmod/ofi/ofi_init.h (hunk 12 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: The original design basically requires the provider to support at 16-bit keys (`max_mr_key_size >= 2 (bytes)`.  Since we use the id space to allocate window id and MR keys for LMT protocol, we want to have a sufficiently large key space. I think this is also related to the Pavan's comment regarding the required OFI feature which we should clarify.

Comment (hajimefu) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: Now that we no longer split the key space, I'd think we can remove the requirement for key size.

Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: Agreed. We shouldn't put arbitrary restrictions on it.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 7 above)
Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: That sounds like an overgeneralization, which should be avoided.  If we really need it later, we can always create a generic version.

File: src/mpid/ch4/netmod/ofi/ofi_comm.h [hunk 13]
```diff
@@ -28,8 +28,8 @@ static inline int MPIDI_NM_mpi_comm_create_hook(MPIR_Comm * comm)
...
-    MPIDI_OFI_index_allocator_create(&MPIDI_OFI_COMM(comm).rma_id_allocator, 1, MPL_MEM_RMA);
+    MPIDI_OFI_index_allocator_create(&MPIDI_OFI_COMM(comm).win_id_allocator, MPL_MEM_RMA);
+    MPIDI_OFI_index_allocator_create(&MPIDI_OFI_COMM(comm).rma_id_allocator, MPL_MEM_RMA);
```
Comment (pavanbalaji) [inline src/mpid/ch4/netmod/ofi/ofi_comm.h]: What is an `rma_id`?  It really needs a better name.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_init.h (hunk 12 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: Fixed in the latest revision

File: src/mpid/ch4/netmod/ofi/ofi_am_impl.h (hunk 11 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_impl.h]: I have removed this check and force pushed.

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 7 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: I have removed the passing of the allocator pointer. Now it is a global variable in specific to the mr_key allocator.

File: src/mpid/ch4/netmod/ofi/ofi_comm.h (hunk 13 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_comm.h]: The `rma_id` is the name for the mr_key used in the LMT protocol. This is removed by a succeeding commit.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_am_events.h (hunk 6 above)
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_am_events.h]: Are we keeping this cast? Why are we not storing the returned key as a `uint64_t`? That is what libfabric defines in the API after all.

Comment (yfguo) [reply src/mpid/ch4/netmod/ofi/ofi_am_events.h]: For this specific patch we do need this cast. However, this will be removed in later commits when key layout is simplified. At the point, the `MPIDI_OFI_rma_key_unpack()` is gone and we will be using 64bit flat key space.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
@raffenet I have cleaned up the casting to `uint64_t`. Can you take another look at the PR?

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 5 above)
Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: I was hoping some explanation would get added to the commit message regarding this change. Another option would be to squash this with the other simplify commit. If that is not trivial because of rebase conflicts, then I'll relent in the interest of getting this merged.

**Comment by yfguo** [comment]
@raffenet Honestly, my stomach cramps when I look at the relationship between commits. I think the reasonable way is to squash very thing into one. I think none of the commits can stand alone without others---they will be either unnecessary (supporting uint64_t) or incorrect (global allocator without key layout simplification).

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
